### PR TITLE
🌱 Add integration test for release notes tool

### PR DESCRIPTION
**Pull request by g-gaston**
**What this PR does / why we need it**:
We make changes to the notes tool pretty frequently, mostly around formatting and handling edge cases. However, the lack of unit tests and the difficulty to add more (given the current structure of the code and the dependency on cli tools) makes it risky to perform these changes with speed. In order to gain confidence on the change, a lot of manual testing is required which is time consuming and prone to errors. This also makes testing for regression difficult.

This PR adds some integration tests (almost e2e): it checks out the right commit, runs the tool and compares the output with a golden file. I decided to test the code directly instead of a built cli to avoid having to build the binary every time we test and also to be able to step with a debugger easily.

This is not perfect since it depends on local git (it uses worktrees to create an "replicable" environment). In addition, it depends on having `gh` with a valid Github authentication, so it won't be trivial to automate. However, it should do the trick for now so folks can run it locally when making changes to the code. If there are changes to the expected output, these will have to be included in the PR, so we can all review them and see if the match the code changes.

/area release
cc @kubernetes-sigs/cluster-api-release-team

### CODE REVIEW

**Review by mjlshen — APPROVED** [review]
/lgtm

File: hack/tools/release/release_notes_integration_test.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	"os"
+	"testing"
+
+	. "github.com/onsi/gomega"
+)
+
+func TestReleaseNotes(t *testing.T) {
+	g := NewWithT(t)
+	folder := os.Getenv("NOTES_TEST_FOLDER")
+	golderFile := os.Getenv("NOTES_TEST_GOLDEN_FILE")
```
Comment (furkatgofurov7) [inline hack/tools/release/release_notes_integration_test.go:34]: nit:

```suggestion
	goldenFile := os.Getenv("NOTES_TEST_GOLDEN_FILE")
```

and in every occurence

File: hack/tools/release/test/run_notes_test.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+script_dir=$(realpath "$(dirname "$0")")
+
+# This tests a patch release computing the PR list from previous patch tag
+# to HEAD. Since v1.3 is out of support, we won't be backporting new PRs
+# so the branch should remain untouched and the test valid.
+test v1.3.9 release-1.3 "$script_dir/golden/v1.3.10.md"
+
+# The release notes command computes everything from last tag
+# to HEAD. Hence if we use the head of release-1.5, this test will
+# become invalid everytime we backport some pr for release cycle v1.5.
```
Comment (furkatgofurov7) [inline hack/tools/release/test/run_notes_test.sh:48]: perhaps you meant

```suggestion
# become invalid everytime we backport some PR to release branch release-1.5.
```
?

**Review by furkatgofurov7 — COMMENTED** [review]
Looks great, thanks!

Nothing major, some minor comments:

**Review by furkatgofurov7 — COMMENTED** [review]
/lgtm

**Review by killianmuldoon — COMMENTED** [review]
Haven't given this a code level review - but a couple of questions.

1 - could we have this entire test in Go instead of using a shell script + the golden files  + a go file?
2 -We should add a `make` target and try to run it as a normal part of our test suite if possible.

**Comment by g-gaston** [comment]
@killianmuldoon 

> Haven't given this a code level review - but a couple of questions.
> 
> 1 - could we have this entire test in Go instead of using a shell script + the golden files + a go file?

I can give that a try. I can probably get rid of the bash script by doing all the setup in go running git commands. However, what's is the issue you see with golden files? We still need something to compare the output with.

> 2 -We should add a `make` target and try to run it as a normal part of our test suite if possible.

The problem with running this in CI is it depends on local git (it uses worktrees to create an "replicable" environment). I'm not sure if the git tree is available when running tests, although we can probably configure it if necessary.

In addition, it depends on having gh with a valid Github authentication, so it won't be trivial to automate.

Given the complexity of automating this, I would prefer if we move forward with this PR and have the CI addition as in improvement for the next release cycle. At least that gives us something to verify changes against in the meantime, even if it has to be run manually. In addition, if we merge #9618, we remove the dependency on git, which should make automating this easier.

File: hack/tools/go.mod [hunk 3]
```diff
@@ -9,6 +9,7 @@ replace sigs.k8s.io/cluster-api/test => ../../test
 require (
 	cloud.google.com/go/storage v1.35.1
 	github.com/blang/semver/v4 v4.0.0
+	github.com/onsi/gomega v1.30.0
```
Comment (g-gaston) [inline hack/tools/go.mod:12]: This is my only real change
The other modules just come from running `go mod tidy`

**Comment by g-gaston** [comment]
@killianmuldoon got rid of the bash script and did everything in the go test file

**Comment by sbueringer** [comment]
> @killianmuldoon got rid of the bash script and did everything in the go test file

Just fyi, Killian is on PTO (IIRC for another 2-3 weeks)

File: hack/tools/release/notes/main.go [hunk 4]
```diff
@@ -143,7 +143,7 @@ const (
 	missingAreaLabelPrefix   = "MISSING_AREA"
 	areaLabelPrefix          = "area/"
 	multipleAreaLabelsPrefix = "MULTIPLE_AREAS["
-	documentationAreaLabel   = "documentation"
+	documentationAreaLabel   = "Documentation"
```
Comment (furkatgofurov7) [inline hack/tools/release/notes/main.go:146]: probably this was taken care already, you might need to rebase the PR

File: hack/tools/release/notes/release_notes_integration_test.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+			g.Expect(err).ToNot(HaveOccurred())
+
+			orgCurrentDir, err := os.Getwd()
+			g.Expect(err).To(Succeed())
+			t.Cleanup(func() {
+				g.Expect(os.Chdir(orgCurrentDir)).To(Succeed())
+			})
+			g.Expect(os.Chdir(tc.releaseBranchForTest)).To(Succeed())
+
+			// two workers is slow but is guarantees no rate limiting
```
Comment (furkatgofurov7) [inline hack/tools/release/notes/release_notes_integration_test.go:77]:
```suggestion
			// two workers are slow but guarantees no rate limiting
```

**Review by furkatgofurov7 — COMMENTED** [review]
Few small nits, otherwise looks good to me.

**Review by furkatgofurov7 — COMMENTED** [review]
/approve

**Review by furkatgofurov7 — COMMENTED** [review]
/lgtm

**Comment by furkatgofurov7** [comment]
cc @CecileRobertMichon @vincepri hey folks 👋🏼 Can you please take a look and help with reviewing/merging this one? Thank you!

**Comment by g-gaston** [comment]
> cc @CecileRobertMichon @vincepri hey folks 👋🏼 Can you please take a look and help with reviewing/merging this one? Thank you!

PS: I think the only reason why Furkat can't approve this one is because I added a new target to the Makefile and change the go.mod (which are both outside of the `release` folder)

**Review by CecileRobertMichon — COMMENTED** [review]
/approve
